src/v/kafka/server/fetch_results_cache.hh

```diff
@@ -0,0 +1,126 @@
+
+namespace kafka {
+
+/**
+ * Simple type aggregating either data or an error
+ */
+struct read_result {
+    using foreign_data_t = ss::foreign_ptr<std::unique_ptr<iobuf>>;
+    using data_t = std::unique_ptr<iobuf>;
+    using variant_t = std::variant<data_t, foreign_data_t>;
+    explicit read_result(error_code e)
+      : error(e) {}
+
+    read_result(
+      variant_t data,
+      model::offset start_offset,
+      model::offset hw,
+      model::offset lso,
+      std::vector<cluster::rm_stm::tx_range> aborted_transactions)
+      : data(std::move(data))
+      , start_offset(start_offset)
+      , high_watermark(hw)
+      , last_stable_offset(lso)
+      , error(error_code::none)
+      , aborted_transactions(std::move(aborted_transactions)) {}
+
+    read_result(model::offset start_offset, model::offset hw, model::offset lso)
+      : start_offset(start_offset)
+      , high_watermark(hw)
+      , last_stable_offset(lso)
+      , error(error_code::none) {}
+
+    bool has_data() const {
+        if (std::holds_alternative<data_t>(data)) {
+            return std::get<data_t>(data) != nullptr;
+        } else {
+            return !std::get<foreign_data_t>(data)->empty();
+        }
+    }
+
+    const iobuf& get_data() const {
+        if (std::holds_alternative<data_t>(data)) {
+            return *std::get<data_t>(data);
+        } else {
+            return *std::get<foreign_data_t>(data);
+        }
+    }
+
+    iobuf release_data() && {
+        if (std::holds_alternative<data_t>(data)) {
+            return std::move(*std::get<data_t>(data));
+        } else {
+            return std::move(*std::get<foreign_data_t>(data));
+        }
+    }
+
+    variant_t data;
+    model::offset start_offset;
+    model::offset high_watermark;
+    model::offset last_stable_offset;
+    model::offset fetch_offset;
+    error_code error;
+    model::partition_id partition;
+    std::vector<cluster::rm_stm::tx_range> aborted_transactions;
+};
+/**
+ * Results cache is a simple data structure that is held in a connection
+ * context. We use it to store across the requests, fetch read results that
+ * would cause fetch response size to larger than allowed fetch.max.bytes.
+ */
+class fetch_results_cache {
+public:
+    explicit fetch_results_cache(size_t);
+    fetch_results_cache(const fetch_results_cache&) = delete;
+    fetch_results_cache(fetch_results_cache&&) = default;
+
+    fetch_results_cache& operator=(const fetch_results_cache&) = delete;
+    fetch_results_cache& operator=(fetch_results_cache&&) = delete;
+
+    ~fetch_results_cache();
+
+    bool put(model::ntp ntp, read_result res);
+    std::optional<read_result> try_get(const model::ntp& ntp, model::offset o);
+
+private:
+    struct entry {
+        explicit entry(read_result r)
+          : result(std::move(r))
+          , timestamp(ss::lowres_clock::now()) {}
+
+        read_result result;
+        ss::lowres_clock::time_point timestamp;
+    };
+
+    void evict();
```

Comment:
should we register the reclaimer here so that it can trigger evictions in low-memory situations?

Reply:
do you think we need reclaimer for this small cache. Using reclaimer would make this cache much more complicated.

Reply:
for how long?

Reply:
wait... i thought we decided to not do this cache and instead to focus on bringing back metadata and computing the next 'Plan' of execution.

Reply:
we agreed that the 'plan' will be a next step, we need some intermediate solution that will allow us to move on

Reply:
all entries are evicted from the cache after 10 seconds

Reply:
are we evicting after a period of time, and _also_ not letting it exceed a total size? a lot of data could accumulate in 10 seconds...

given we are looking for short term solutions that aren't computing the execution plan, another option worth considering for the short term solution space is to use a global structure on the heap that is shared between cores but there are no atomics so it'd be like a sloppy accumulator. just a thought.

Reply:
I've replaced cache with an execution plan.